YatimaStdLib/GaloisField.lean

```diff
@@ -0,0 +1,206 @@
+import YatimaStdLib.Polynomial
+import YatimaStdLib.Zmod
+import YatimaStdLib.Bit
+
+/-!
+# Galois Fields
+
+This module provides the basic data structures necessary to define and work with prime fields
+and their extensions.
+
+Here we post some definitions from https://hackage.haskell.org/package/galois-field-1.0.2
+-/
+
+/-- The structure of a Galois field on t-/
+class GaloisField (K : Type _) where
+  plus : K → K → K
+  times : K → K → K
+  null : K
+  ein : K
+  minus : K → K → K
+  divis : K → K → K
+  -- Characteristic `p` of field and order of prime subfield.
+  char : Nat
+  -- Degree `q` of field as extension field over prime subfield.
+  deg : Nat
+  -- Frobenius endomorphism `x ↦ x^p` of prime subfield.
+  frob : K → K
+
+namespace GaloisField
+
+instance [GaloisField K] : Inhabited K where
+  default := null
+
+instance [GaloisField K] : Add K where
+  add := plus
+
+instance [GaloisField K] : Mul K where
+  mul := times
+
+instance [GaloisField K] : OfNat K (nat_lit 1) where
+  ofNat := ein
+
+instance [GaloisField K] : OfNat K (nat_lit 0) where
+  ofNat := null
+
+instance [GaloisField K] : Div K where
+  div := divis
+
+instance [GaloisField K] : Sub K where
+  sub := minus
+
+/-- An `O(log n)` implementation of `galPow` -/
+def fastPow [Mul K] [OfNat K (nat_lit 1)] (x : K) (n : Nat) : K := 
+  let binExp := n.toBits
+  let squares := getSquares [x] (binExp.length)
+  binExp.zip squares |>.foldl (init := 1) (fun acc (x, s) => if x == .one then acc * s else acc)
+  where getSquares (acc : List K) (n : Nat) : List K :=
+    match n with
+    | 0 => []
+    | 1 => acc
+    | n + 1 => getSquares ((acc.headD x) * (acc.headD x) :: acc) n
+
+def galPow [GaloisField K] : K → Nat → K
+  | _, 0 => 1
+  | x, (k + 1) => x * (galPow x k)
+
+-- TODO: Replace this with `fastPow`?
+instance [GaloisField K] : HPow K Nat K where
+  hPow := galPow
```

Comment:
`galPow` actually already utilises `galPow`, so the comment above doesn't seem needed

Reply:
The question was whether to replace `galPow` with `fastPow` in the `HPow` instance for `K`... I think I've decided to keep it as is, and instead offer `fastPow` as an alternative when we need it (for example, computing the Frobenius in a field of size ~2^256, `fastPow` is almost 100% required)